LiS/man/man9/cmn_err.9.man
'\" rtp
.\" -*- nroff -*- vim: ft=nroff nocin nosi
.\"
.\" @(#) $Id: cmn_err.9.man,v 1.1.4.6 2008-04-29 08:33:21 brian Exp $
.\"
.\" =========================================================================
.\"
.\" Copyright (c) 2001-2008  OpenSS7 Corporation <http://www.openss7.com/>
.\"
.\" All Rights Reserved.
.\"
.\" Permission is granted to make and distribute verbatim copies of this
.\" manual provided the copyright notice and this permission notice are
.\" preserved on all copies.
.\"
.\" Permission is granted to copy and distribute modified versions of this
.\" manual under the conditions for verbatim copying, provided that the
.\" entire resulting derived work is distributed under the terms of a
.\" permission notice identical to this one
.\" 
.\" Since the Linux kernel and libraries are constantly changing, this
.\" manual page may be incorrect or out-of-date.  The author(s) assume no
.\" responsibility for errors or omissions, or for damages resulting from
.\" the use of the information contained herein.  The author(s) may not
.\" have taken the same level of care in the production of this manual,
.\" which is licensed free of charge, as they might when working
.\" professionally.  The author(s) will take no responsibility in it.
.\" 
.\" Formatted or processed versions of this manual, if unaccompanied by
.\" the source, must acknowledge the copyright and authors of this work.
.\"
.\" -------------------------------------------------------------------------
.\"
.\" U.S. GOVERNMENT RESTRICTED RIGHTS.  If you are licensing this Software
.\" on behalf of the U.S. Government ("Government"), the following
.\" provisions apply to you.  If the Software is supplied by the Department
.\" of Defense ("DoD"), it is classified as "Commercial Computer Software"
.\" under paragraph 252.227-7014 of the DoD Supplement to the Federal
.\" Acquisition Regulations ("DFARS") (or any successor regulations) and the
.\" Government is acquiring only the license rights granted herein (the
.\" license rights customarily provided to non-Government users).  If the
.\" Software is supplied to any unit or agency of the Government other than
.\" DoD, it is classified as "Restricted Computer Software" and the
.\" Government's rights in the Software are defined in paragraph 52.227-19
.\" of the Federal Acquisition Regulations ("FAR") (or any successor
.\" regulations) or, in the cases of NASA, in paragraph 18.52.227-86 of the
.\" NASA Supplement to the FAR (or any successor regulations).
.\"
.\" =========================================================================
.\" 
.\" Commercial licensing and support of this software is available from
.\" OpenSS7 Corporation at a fee.  See http://www.openss7.com/
.\" 
.\" =========================================================================
.\"
.\" Last Modified $Date: 2008-04-29 08:33:21 $ by $Author: brian $
.\"
.\" -------------------------------------------------------------------------
.\"
.\" $Log: cmn_err.9.man,v $
.\" Revision 1.1.4.6  2008-04-29 08:33:21  brian
.\" - update headers for Affero release
.\"
.\" Revision 1.1.4.5  2006/04/05 09:42:43  brian
.\" - moved .so line and updated headers
.\"
.\" =========================================================================
.R1
bracket-label "\fR[\fB" "\fR]" "\fR, \fB"
no-default-database
database lis.refs
accumulate
move-punctuation
abbreviate A
join-authors ", " ", " " and "
et-al " et al" 2 3
abbreviate-label-ranges ".."
sort-adjacent-labels
.R2
.so lis.macros
.\"
.\"
.TH CMN_ERR 9 "@PACKAGE_DATE@" "@PACKAGE@-@VERSION@" "@PACKAGE_TITLE@ DDI/DKI"
.\"
.\"
.SH NAME
.B cmn_err, lis_cmn_err
\- print a kernel command error
.\"
.\"
.SH SYNOPSIS
.PP
.B #include <sys/cmn_err.h>
.HP 8
.BI "cmn_err(int " err_lvl ", char *" fmt ", " ... );
.\"
.\"
.SH ARGUMENTS
.PP
.TP
.I err_lvl
the error level to report.
.TP
.I fmt
a
.BR printf (3)
format string.
.TP
.I ...
the arguments to complete the format string.
.\"
.\"
.SH DESCRIPTION
.PP
.BR cmn_err ()
prints a command error message to kernel log files at the level specified by
.I err_lev
and with the
.BR printk ()
formatted format string
.I fmt
and associated variable arguments.
.PP
.I err_lvl
may be one of the following values:
.TP 12
.B CE_CONT
This level indicates to continue printing.
.TP 12
.B CE_NOTE
This error level is for notices.  Notices are important events which should be
logged.
This equates to a
.B KERN_NOTICE
level in
.BR Linux .
.TP 12
.B CE_WARN
This error level is for warinings.  These are critical conditions, however,
these errors do not halt the kernel.
This equates to a
.B KERN_WARNING
level in
.BR Linux .
.TP 12
.B CE_PANIC
This error level is a panic error level.  This is for highly critical errors
where the kernel has destabilized and it is necessary to halt the kernel.
.RS
.PP
When the
.I err_lvl
is specified as
.BR CE_PANIC ,
.BR cmn_err ()
will generate a
.B Linux
kernel panic with the resulting message.
.RE
.\"
.\"
.SH RETURN
.PP
Unlike
.BR printk (),
.BR cmn_err ()
returns
.BR void .
.\"
.\"
.SH ERRORS
.PP
.BR cmn_err ()
fails silently when the
.I err_lvl
is invalid or formatting of the message
per
.I fmt
and the variable arguments fails.
.\"
.\"
.SH CONTEXT
.PP
.BR cmn_err ()
can be called from any context, including user context, module procedures,
callouts, callbacks, soft interrupts and interrupt service routines.
.\"
.\"
.SH MP-STREAMS
.PP
.BR cmn_err ()
is MP-safe.
.\"
.\"
.SH NOTICES
.PP
.B LiS
.[
LiS
.]
uses
.BR printk ()
to print the messages.
.\"
.\"
.SH "SEE ALSO"
.PP
.BR printk ().
.\"
.\"
.SH BUGS
.PP
.BR cmn_err ()
has no known bugs.
.\"
.\"
.SH COMPATIBILITY
.PP
.BR cmn_err ()
is compatible with
.IR "SVR 4.2 MP DDI/DKI" ,
.[
usldk
.]
and implementations based on
.IR "SVR 4" ,
.[
svr4dk
.]
with the following portability considerations:
.IP \(em \w'\(em\(em'u
SVR 4 systems normally print a newline at before the information provided to
.BR cmn_err ()
unless the
.B CE_CONT
flag is specified, and do not normally print a newline after the information provided to
.BR cmn_err ().
This is the opposite of BSD and Linux behavior, which print a newline at the end of the sequence
rather than the beginning.
.B LiS
can be compiled with flags providing either behavior.
.PP
For a fully functional
.I STREAMS
implementation, use
.BR "Linux Fast-STREAMS" .
.[
LfS
.]
.\"
.\"
.SH CONFORMANCE
.PP
SVR 4.2 MP DDI/DKI.
.[
usldk
.]
.\"
.\"
.SH HISTORY
.PP
.BR cmn_err ()
is a System V function that first appeared in SVR 2.
.\"
.\"
.[
$LIST$
.]
.TI


NEWS
OpenSS7 Master Package -- history of user visible changes.  2008-05-03
$Id: NEWS,v 0.9.2.53 2008-05-03 21:22:25 brian Exp $
Copyright (c) 2001-2008  OpenSS7 Corporation. <http://www.openss7.com/>
Copyright (c) 1997-2000  Brian Bidulock <bidulock@openss7.org>
See the end for copying conditions (for this file).

The sections that follow provide information on OpenSS7 releases of the
OpenSS7 Master Package package.

 Major changes for release openss7-0.9.2.G
------------------------------------------

This is the sixth public release of the OpenSS7 Master Package.  See
`README' in the release for a sub-package listing.  Most of the
sub-packages in the release are production grade for `Linux
Fast-STREAMS'.  All existing validation test suites run clean on
supported distributions and architectures.

The `OpenSS7 Master Package' is not released as often as the
sub-packages.  As sub-packages are released more often, to rebuild the
master package with a new sub-package release, simply replace the
directory to which the sub-package belongs with the unpacked
sub-package release and then rebuild the master package.  This release
provides support for recent distributions and tool chains.

Major features since the last public release are as follows:

   - Documented `lib32gcc1' problem on Ubuntu.

   - Noted problem running under XEN kernels.  XEN kernels are not yet
     supported.

   - License upgrade to AGPL v3.

   - Cross-compile fixes (strap out AC_FUNC_REALLOC macro when
     cross-compiling).

   - Ability to strap out major documentation build and installation
     primarily for embedded targets.

   - Improvements to common build process for embedded and
     cross-compile targets.

   - Support for flex 2.5.33  in maintainer mode.

   - Added the following sub-packages to the master build (and relase):

         - `strx25-0.9.2.1'


_This is a public stable production grade release of the package: it
deprecates previous releases.  Please upgrade to the current release
before reporting bugs._

As with other OpenSS7 releases, this release configures, compiles,
installs and builds RPMs and DEBs for a wide range of Linux 2.4 and 2.6
RPM- and DPKG-based distributions, and can be used on production
kernels without patching or recompiling the kernel.

This package is publicly released under the `GNU Affero General Public
License Version 3'.  The release is available as an `autoconf' tarball,
SRPM, DSC, and set of binary RPMs and DEBs.  See the downloads page
(http://www.openss7.org/download.html) for the `autoconf' tarballs,
SRPMs and DSCs.  For tarballs, SRPMs, DSCs and binary RPMs and DEBs,
see the openss7 package page (http://www.openss7.org/openss7_pkg.html).

See `http://www.openss7.org/codefiles/openss7-0.9.2.G/ChangeLog' and
`http://www.openss7.org/codefiles/openss7-0.9.2.G/NEWS' in the release
for more information.  Also, see the `openss7.pdf' manual in the
release (also in html `http://www.openss7.org/openss7_manual.html').

 Major changes for release openss7-0.9.2.F
------------------------------------------

This is the fifth public release of the OpenSS7 Master Package.  See
`README' in the release for a sub-package listing.  Most of the
sub-packages in the release are production grade for `Linux
Fast-STREAMS'.  All existing validation test suites run clean on
supported distributions and architectures.

The `OpenSS7 Master Package' is not released as often as the
sub-packages.  As sub-packages are released more often, to rebuild the
master package with a new sub-package release, simply replace the
directory to which the sub-package belongs with the unpacked
sub-package release and then rebuild the master package.  This release
provides support for recent distributions and tool chains.

Major features since the last public release are as follows:

   - Support build on openSUSE 10.2.

   - Support build on Fedora 7 and 2.6.21 kernel.

   - Support build on CentOS 5.0 (RHEL5).

   - Support build on Ubuntu 7.04.

   - Updated to gettext 0.16.1.

   - Changes to support build on 2.6.20-1.2307.fc5 and
     2.6.20-1.2933.fc6 kernel.

   - Supports build on Fedora Core 6.

   - Support for recent distributions and tool chains.


 Major changes for release openss7-0.9.2.E
------------------------------------------

This is the fourth public release of the OpenSS7 Master Package.  See
`README' in the release for a sub-package listing.  Most of the
sub-packages in the release are production grade for `Linux
Fast-STREAMS'.  All existing validation test suites run clean on
supported distributions and architectures.

It is unlikely that the `OpenSS7 Master Package' will be released as
frequently as before.  Sub-packages will be released more often.  To
rebuild the master package with a new sub-package release, simply
replace the directory to which the sub-package belongs with the
unpacked sub-package release and then rebuild the master package.  This
release provides support for recent distributions and tool chains.

Major features since the last public release are as follows:

   - Addition of the `osr61' sub-package that contains Dialogic(R) Open
     System Release 6.1 version 239 GPL drivers.

   - A few minor corrections to the common build process.

   - Support for autoconf 2.61, automake 1.10 and gettext 0.16.

   - Support for Ubuntu 6.10 distribution and bug fixes for i386
     kernels.

   - The package now looks for subpackages with a version number as
     unpacked by separate tarball.


 Major changes for release openss7-0.9.2.D
------------------------------------------

This is the fourth public release of the OpenSS7 Master Package.  The
sub-packages have been reorganized for this release.  See `README' in
the release for a sub-package listing.  Aside from sub-package
reorganization, the major difference from previous release is that this
release no longer contains `LiS'.  Too many of the sub-packages will
not even build against `LiS' because of its Stream head deficiencies.


LiS/head/cmn_err.c
/*****************************************************************************

 @(#) $RCSfile: cmn_err.c,v $ $Name:  $($Revision: 1.1.1.2.4.9 $) $Date: 2008-04-29 08:33:12 $

 -----------------------------------------------------------------------------

 Copyright (c) 2001-2008  OpenSS7 Corporation <http://www.openss7.com/>
 Copyright (c) 1997-2000  Brian F. G. Bidulock <bidulock@openss7.org>

 All Rights Reserved.

 This program is free software: you can redistribute it and/or modify it under
 the terms of the GNU General Public License as published by the Free Software
 Foundation, version 3 of the license.

 This program is distributed in the hope that it will be useful, but WITHOUT
 ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
 FOR A PARTICULAR PURPOSE.  See the GNU General Public License for more
 details.

 You should have received a copy of the GNU General Public License along with
 this program.  If not, see <http://www.gnu.org/licenses/>, or write to the
 Free Software Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.

 -----------------------------------------------------------------------------

 U.S. GOVERNMENT RESTRICTED RIGHTS.  If you are licensing this Software on
 behalf of the U.S. Government ("Government"), the following provisions apply
 to you.  If the Software is supplied by the Department of Defense ("DoD"), it
 is classified as "Commercial Computer Software" under paragraph 252.227-7014
 of the DoD Supplement to the Federal Acquisition Regulations ("DFARS") (or any
 successor regulations) and the Government is acquiring only the license rights
 granted herein (the license rights customarily provided to non-Government
 users).  If the Software is supplied to any unit or agency of the Government
 other than DoD, it is classified as "Restricted Computer Software" and the
 Government's rights in the Software are defined in paragraph 52.227-19 of the
 Federal Acquisition Regulations ("FAR") (or any successor regulations) or, in
 the cases of NASA, in paragraph 18.52.227-86 of the NASA Supplement to the FAR
 (or any successor regulations).

 -----------------------------------------------------------------------------

 Commercial licensing and support of this software is available from OpenSS7
 Corporation at a fee.  See http://www.openss7.com/

 -----------------------------------------------------------------------------

 Last Modified $Date: 2008-04-29 08:33:12 $ by $Author: brian $

 -----------------------------------------------------------------------------

 $Log: cmn_err.c,v $
 Revision 1.1.1.2.4.9  2008-04-29 08:33:12  brian
 - update headers for Affero release

 Revision 1.1.1.2.4.8  2007/08/14 10:47:00  brian
 - GPLv3 header update

 *****************************************************************************/

#ident "@(#) $RCSfile: cmn_err.c,v $ $Name:  $($Revision: 1.1.1.2.4.9 $) $Date: 2008-04-29 08:33:12 $"

static char const ident[] = "$RCSfile: cmn_err.c,v $ $Name:  $($Revision: 1.1.1.2.4.9 $) $Date: 2008-04-29 08:33:12 $";

/************************************************************************
*                           cmn_err                                     *
*************************************************************************
*									*
* This is the UNIX common-error message routine.			*
*									*
* For now we will simply turn this into a printk().  Later we can	*
* add some putbuf code.							*
*									*
* Author:	David Grothe	<dave@gcom.com>				*
*									*
* Copyright (C) 1997  David Grothe, Gcom, Inc <dave@gcom.com>		*
*									*
************************************************************************/

#include <sys/strport.h>
#include <sys/cmn_err.h>
#include <stdarg.h>
#include <sys/osif.h>

/*
 * Check to see if the symbols for kernel message level have
 * been defined.  If so, use them, if not use null string
 * defaults.  In the case that we are compiled into the Linux
 * kernel this will have the effect of mapping cmn_err message
 * levels to Linux kernel message priorities.  For other
 * environments this will not apply.
 *
 * If we are in Solaris compatible mode then omit the kernel
 * message identifiers.  In Solaris mode the newline goes at the
 * end rather than at the beginning so the kernel flags might
 * be misinterpreted.
 *
 * For SVR4 the newline goes first so it is safe to interpolate the
 * kernel message type tags.
 */

#if defined(SOLARIS_STYLE_CMN_ERR)	/* config option */

#define	PRE_NL		""
#undef  KERN_NOTICE
#define	KERN_NOTICE	""
#undef  KERN_WARNING
#define	KERN_WARNING	""
#undef  KERN_EMERG
#define	KERN_EMERG	""

#else				/* SVR4 style cmn_err */

#define	PRE_NL		"\n"
#ifndef KERN_NOTICE
#define	KERN_NOTICE	""
#endif
#ifndef KERN_WARNING
#define	KERN_WARNING	""
#endif
#ifndef KERN_EMERG
#define	KERN_EMERG	""
#endif
#ifndef KERN_INFO
#define	KERN_INFO	""
#endif

#endif

lis_spin_lock_t lis_cmn_err_lock;
char lis_cmn_err_buf[4096];

#define	buf lis_cmn_err_buf

void
lis_cmn_err_init(void)
{
	lis_spin_lock_init(&lis_cmn_err_lock, "CmnErr-Lock");
}

void _RP
lis_vcmn_err(int err_lvl, const char *fmt, va_list args)
{
	lis_flags_t psw;
	char *p;

	lis_spin_lock_irqsave(&lis_cmn_err_lock, &psw);
	switch (err_lvl) {
	case CE_CONT:		/* continue printing */
		buf[0] = 0;
		break;
	case CE_NOTE:		/* NOTICE */
		strcpy(buf, PRE_NL KERN_NOTICE "NOTICE: ");
		break;
	case CE_WARN:		/* WARNING */
		strcpy(buf, PRE_NL KERN_WARNING "WARNING: ");
		break;
	case CE_PANIC:		/* PANIC */
		strcpy(buf, PRE_NL KERN_EMERG "PANIC: ");
		break;
	default:
		lis_spin_unlock_irqrestore(&lis_cmn_err_lock, &psw);
		printk("\n" KERN_NOTICE "cmn_err:  Called with invalid arguments "
		       "(0x%lx, 0x%lx)\n", (long) err_lvl, (long) fmt);
		return;
	}

	for (p = buf; *p; p++) ;	/* find end of string */

	vsprintf(p, fmt, args);

#if defined(SOLARIS_STYLE_CMN_ERR)	/* config option */
	if (err_lvl != CE_CONT) {
		for (; *p; p++) ;	/* find end of string */
		*p++ = '\n';	/* add newline at end */
		*p = 0;
	}
#endif

	if (err_lvl == CE_PANIC)
		PANIC(buf);
	else
		printk("%s", buf);

	lis_spin_unlock_irqrestore(&lis_cmn_err_lock, &psw);

}				/* lis_cmn_err */

void _RP
lis_cmn_err(int err_lvl, const char *fmt, ...)
{
	va_list args;

	va_start(args, fmt);
	lis_vcmn_err(err_lvl, fmt, args);
	va_end(args);
	return;

}				/* lis_cmn_err */
